### PR TITLE
added correct interrupt handler initialization in RISC-V-Qemu-virt_GCC

### PR DESCRIPTION
**Pull request by FriedDede**
Corrected interrupt handler initialization in RISC-V-Qemu-virt_GCC
-----------

Description
-----------
Added asm to load the correct interrupt handler in mtvec register

Steps
-----------
The demo as in the main branch is not working because it lacks interrupt handler initialization.
Compile and run the demo as in README to reproduce the issue.
Then try to run the demo with my addition to verify the fix.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by kstribrnAmzn** [comment]
Thanks for submitting the PR @FriedDede! I'm going to look over this PR now. I'll post any comments/questions on the bits of code.

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c [hunk 1]
```diff
@@ -30,6 +30,10 @@
 
 /* Run a simple demo just prints 'Blink' */
 #define DEMO_BLINKY	1
+#define mainVECTOR_MODE_DIRECT	1
+
+extern void freertos_risc_v_trap_handler( void );
+extern void freertos_vector_table( void );
```
Comment (kstribrnAmzn) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:36]: Where is your `freertos_vector_table` coming from? It looks typical to define this vector_table in an assembly file (see [this file for example](https://github.com/FreeRTOS/FreeRTOS/blob/4255ac5c28af626ea7953006a4ff112f758ea332/FreeRTOS/Demo/RISC-V_RV32_QEMU_VIRT_GCC/vector.S)).

Comment (FriedDede) [reply FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:36]: You are right, I forgot to push the assembly file, i was focused to get the demo working in direct mode and forgot to push the file with the vector definition. 
Sorry for my mistake, how can I fix that?

Comment (kstribrnAmzn) [reply FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:36]: No sweat! Should be as simple as making another commit with the assembly file and pushing it to this PR.

Comment (FriedDede) [reply FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:36]: now is ok, sorry for the mess, never done a PR before

Comment (FriedDede) [reply FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:36]: The fix is been actually taken from the example you posted btw, so it should work in the same way.

Comment (kstribrnAmzn) [reply FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:36]: >sorry for the mess, never done a PR before

No need to be sorry. Having multiple commits for a PR is basically standard practice. This isn't a mess.

Comment (kstribrnAmzn) [reply FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:36]: Awesome! Thanks for the quick update. I'm trying to build this demo. Crosstools-ng doesn't like my Mac so I'm trying on a linux machine.

**Comment by kstribrnAmzn** [comment]
I was finally able to get a successful build on my end for the unmodified code. Were you able to build on your end? Did you have to modify the Makefile or differ from the instructions in any way?

I ask because the instructions in the README required a little tweaking for Ubuntu systems. If you were on a different system I'd love to know if these tweaks apply broadly or are on a per-system basis.

**Comment by FriedDede** [comment]
> I was finally able to get a successful build on my end for the unmodified code. Were you able to build on your end? Did you have to modify the Makefile or differ from the instructions in any way?
> 
> I ask because the instructions in the README required a little tweaking for Ubuntu systems. If you were on a different system I'd love to know if these tweaks apply broadly or are on a per-system basis.

I'm on latest fedora, i didn't use crostools-ng but i compiled the toolchain from [here](https://github.com/riscv-collab/riscv-gnu-toolchain) with these commands:
`./configure --prefix=/opt/riscv --with-arch=rv32gc --with-abi=ilp32d
make`

the compiled toolchain files begins with riscv32 instead of riscv64 as in the makefile, so i had to change that.

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c (hunk 1 above)
Comment (FriedDede) [reply FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:36]: Hi, did you manage to test the fix?

**Comment by kstribrnAmzn** [comment]
Sorry for the lack of an update @FriedDede. I didn't have time to further test this until today. I've been working on testing this.

I must be doing something wrong when trying to test (and trying to figure this out now). GDB jumps to a seemingly bad address when debugging the QEMU demo.

My frame when starting the debugger looks like `#0  0x00001000 in ?? ()`  and upon running `continue` or `next` I'm receiving the output `Cannot find bounds of current function`.

________
I'll circle back when I have this figured out

**Comment by FriedDede** [comment]
Are u talking about the Demo without my changes?

**Comment by kstribrnAmzn** [comment]
> Are u talking about the Demo without my changes?

With or without - doesn't seem to matter.

**Comment by FriedDede** [comment]
> > Are u talking about the Demo without my changes?
> 
> With or without - doesn't seem to matter.

The bug that I encountered was that after the startScheduler function call in main_blinky.c the demo jumped to a bad assembly address, that was because the mtvec was uninitialized so the first interrupt went nowhere, but the demo was working correctly before that function call.

I've ran it as written in the readme so it's really strange it isn't working on your end.

**Comment by kstribrnAmzn** [comment]
Just want to keep you up to date on my weirdness. I was able to _finally_ run the vanilla demo. Took a whole bunch of debugging which didn't really pan out. You're statement of...

>The bug that I encountered was that after the startScheduler function call in main_blinky.c the demo jumped to a bad assembly address, that was because the mtvec was uninitialized so the first interrupt went nowhere, but the demo was working correctly before that function call.

is exactly what happens. 

___________

As for me blocking you - I'm sorry about this. I've roped in another teammate so we can hopefully get this demo running flawlessly on his machine. I can't see any issues with your code and your changes are not responsible for the demo failures I've been receiving, but I'd still like to run the demo (even if vicariously) before approving.

The issue I was seeing is that the jump-and-link instruction (aka `jal`) found here is failing to jump to the main function. This makes absolutely no sense as the main function address is correctly defined and linked. I've checked a couple other things and the main function is there - it's just breaking on the jump for a reason unknown to me. Even more frustrating is that last night after trying for like the 20th time on the unmodified demo, it worked, however it went right back to failing on all subsequent runs. 

I wonder if the problem is with the emulator or the specific micro arch that I'm emulating.

**Comment by FriedDede** [comment]
`>` Just want to keep you up to date on my weirdness. I was able to _finally_ run the vanilla demo. Took a whole bunch of debugging which didn't really pan out. You're statement of...
> 
> > The bug that I encountered was that after the startScheduler function call in main_blinky.c the demo jumped to a bad assembly address, that was because the mtvec was uninitialized so the first interrupt went nowhere, but the demo was working correctly before that function call.
> 
> is exactly what happens.
> 
> As for me blocking you - I'm sorry about this. I've roped in another teammate so we can hopefully get this demo running flawlessly on his machine. I can't see any issues with your code and your changes are not responsible for the demo failures I've been receiving, but I'd still like to run the demo (even if vicariously) before approving.
> 
> The issue I was seeing is that the jump-and-link instruction (aka `jal`) found here is failing to jump to the main function. This makes absolutely no sense as the main function address is correctly defined and linked. I've checked a couple other things and the main function is there - it's just breaking on the jump for a reason unknown to me. Even more frustrating is that last night after trying for like the 20th time on the unmodified demo, it worked, however it went right back to failing on all subsequent runs.
> 
> I wonder if the problem is with the emulator or the specific micro arch that I'm emulating.

I understand your concerns, those are my setup specs, so you can have a complete view on what i have on my side.
qemu: ` QEMU emulator version 6.2.0 (qemu-6.2.0-15.fc36) `
Riscv32-gcc : 
` Using built-in specs.
COLLECT_GCC=riscv32-unknown-elf-gcc
COLLECT_LTO_WRAPPER=/home/andre/src/AOS-Project/toolchain/bin/../libexec/gcc/riscv32-unknown-elf/12.2.0/lto-wrapper
Target: riscv32-unknown-elf
Configured with: /home/andre/src/RISCV/riscv-gnu-toolchain/gcc/configure --target=riscv32-unknown-elf --prefix=/home/andre/src/RISCV/toolchain --disable-shared --disable-threads --enable-languages=c,c++ --with-pkgversion=g2ee5e430018 --with-system-zlib --enable-tls --with-newlib --with-sysroot=/home/andre/src/RISCV/toolchain/riscv32-unknown-elf --with-native-system-header-dir=/include --disable-libmudflap --disable-libssp --disable-libquadmath --disable-libgomp --disable-nls --disable-tm-clone-registry --src=/home/andre/src/RISCV/riscv-gnu-toolchain/gcc --disable-multilib --with-abi=ilp32 --with-arch=rv32gc --with-tune=rocket --with-isa-spec=2.2 'CFLAGS_FOR_TARGET=-Os   -mcmodel=medlow' 'CXXFLAGS_FOR_TARGET=-Os   -mcmodel=medlow'
Thread model: single
Supported LTO compression algorithms: zlib
gcc version 12.2.0 (g2ee5e430018) 
`
Riscv32-gdb 10.1

Let me know if you made progress in debugging it.

**Comment by FriedDede** [comment]
Hi, how is it going with the review? Also I'm not understanding what part of my PR is failing unit testing and how to fix it, do you have any clue?

**Comment by kstribrnAmzn** [comment]
Thank you for your patience and providing so much of your setup. After a lot of debugging (I must have rebuilt like half a dozen variations) I switched over to SiFives prebuilt toolchain and found the demo with your change to work flawlessly.

I'm going to approve and will forward onto our PRBR for approval.

**Comment by kstribrnAmzn** [comment]
Our timing of posting was incredible haha

**Review by kstribrnAmzn — DISMISSED** [review]

**Comment by FriedDede** [comment]
> Our timing of posting was incredible haha - 10 seconds apart

Thank u for your work. Hahah yes, very funny coincidence 🤣

**Comment by FriedDede** [comment]
> Update: Going to push a single commit to your PR to fix our 'core-checker'. It's just a simple URL change in our license.

Ok, thanks for the fix

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c [hunk 2]
```diff
@@ -30,6 +30,10 @@
 
 /* Run a simple demo just prints 'Blink' */
 #define DEMO_BLINKY	1
+#define mainVECTOR_MODE_DIRECT	1
+
+extern void freertos_risc_v_trap_handler( void );
```
Comment (AniruddhaKanhere) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:35]: Can you help me find where is this defined? I cannot seem to find it.

Or is this supposed to be defined by the application as a 'hook'?

**Review by AniruddhaKanhere — COMMENTED** [review]
Thank you for this PR @FriedDede. 

I have a question which if you could clarify would be quite helpful. Maybe @kstribrnAmzn knows the answer as well.

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c (hunk 2 above)
Comment (FriedDede) [reply FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:35]: The interrupt_handler is defined kernel side. You can find it in: `FreeRTOS-Kernel/portable/GCC/riscV/portASM.S`

**Review by aggarg — APPROVED** [review]

**Review by shubnil — APPROVED** [review]

**Review by amazonKamath — APPROVED** [review]

**Comment by FriedDede** [comment]
> The headers seem to be of 2020. Especially vector.S being a new file , should it be having 2022 headers
What do i have to change?

**Review by shubnil — APPROVED** [review]
Looks like the date change will require some larger repo wide changes which can be taken up later.

**Comment by aggarg** [comment]
Thank you for your contribution.

**Pull request merged**